src/guide/3-behaviour-definition/4-agent-communication.rst

```diff
@@ -5,13 +5,195 @@ Communication between agents in FLAMEGPU2 is handled through messages. Messages
+
+.. tabs::
+    
+    .. code-tab:: python
+      
+      # Specify that the "outputdata" agent function has an interaction radius of 2.0
+      outputdata.setRadius(2.0)
+
+    .. code-tab:: cpp
+      
+      // Specify that the "outputdata" agent function has an interaction radius of 2.0f
+      outputdata.setMessageOutput(2.0f);
 
 Reading Messages
 ----------------
 
+Reading a message is very similar to sending one. The second argument in the agent function definition defines the input message communication strategy.
+
+.. tabs::
+
+    .. code-tab:: cpp
+
+      // Define an agent function, "inputdata" which has accepts an input message using the "MsgBruteForce" communication strategy and inputs no messages
+      FLAMEGPU_AGENT_FUNCTION(inputdata, MsgBruteForce, MsgNone) {
+        // Agent function code goes here
+        ...
+      }
+
+The input message type is specified using the ``setMessageInput`` method of the ``AgentFunctionDescription`` object:
+
+
+.. tabs::
+    
+    .. code-tab:: python
+      
+      # Specify that the "inputdata" agent function inputs a "location_message"
+      inputdata.setMessageInput("location_message")
+
+    .. code-tab:: cpp
+      
+      // Specify that the "inputdata" agent function inputs a "location_message"
+      inputdata.setMessageInput("location_message");
+
+With the input message type specified, the message list will be available in the agent function definition. The message list can be iterated over to access each message:
+
+
+.. tabs::
+
+    .. code-tab:: cpp
+
+      // Define an agent function, "inputdata" which has accepts an input message using the "MsgBruteForce" communication strategy and inputs no messages
+      FLAMEGPU_AGENT_FUNCTION(inputdata, MsgBruteForce, MsgNone) {
+        // For each message in the message list
+        for (const auto& message : FLAMEGPU->message_in) {
+          int idFromMessage = message->getVariable<int>("id");
+        }
+      }
+
+**Spatial Messaging**
+If you are using one of the spatial messaging strategies, you will also need to supply the x and y coordinates of this agent to access the relevant messages:
+
+.. tabs::
+
+    .. code-tab:: cpp
+
+      // Define an agent function, "inputdata" which has accepts an input message using the "MsgSpatial" communication strategy and inputs no messages
+      FLAMEGPU_AGENT_FUNCTION(inputdata, MsgSpatial, MsgNone) {
+        // Get this agent's x, y, z variables
+        const float x = FLAMEGPU->getVariable<float>("x");
+        const float y = FLAMEGPU->getVariable<float>("y");
+        const float z = FLAMEGPU->getVariable<float>("z");
+        
+        // For each message in the message list which was output by a nearby agent
+        for (const auto& message : FLAMEGPU->message_in(x, y, z)) {
+          int idFromMessage = message->getVariable<int>("id");
+        }
+      }
+
+**TODO** also needs manual radius check
```

Comment:
Possibly worth giving an example of a manually defined hash function for use with `MsgBucket`, to demonstrate how it can be used to implement more complex forms of communication.